File: python/src/a2_abstract_factory.py
"""
The intent is to provide an interface for creating families of related or dependent
objects without specifying their concrete classes.

Key participants:
- AbstractFactory: Declares an interface for creating each abstract product
- ConcreteFactory: Implements operations to create concrete products
- AbstractProduct: Declares an interface for a type of product
- ConcreteProduct: Defines a product object to be created by the corresponding factory
- Client: Uses only interfaces declared by AbstractFactory and AbstractProduct

Key benefits:
- Isolates concrete classes
- Makes exchanging product families easy
- Promotes consistency among products

Real-world example: A GUI framework that creates widgets (buttons, checkboxes)
for different operating systems (Windows, Mac) ensuring all widgets match the OS style.
"""
from abc import ABC, abstractmethod
from typing import Protocol


# ----- Abstract Products -----
class Button(Protocol):
    def paint(self) -> str:
        ...


class Checkbox(Protocol):
    def paint(self) -> str:
        ...


# ----- Concrete Products (Windows) -----
class WindowsButton:
    def paint(self) -> str:
        return "Windows Button: rectangular, gray background"


class WindowsCheckbox:
    def paint(self) -> str:
        return "Windows Checkbox: square with 'X' mark"


# ----- Concrete Products (Mac) -----
class MacButton:
    def paint(self) -> str:
        return "Mac Button: rounded corners, blue accent"


class MacCheckbox:
    def paint(self) -> str:
        return "Mac Checkbox: circular with checkmark"


# ----- Abstract Factory -----
class GUIFactory(ABC):
    @abstractmethod
    def create_button(self) -> Button:
        pass

    @abstractmethod
    def create_checkbox(self) -> Checkbox:
        pass


# ----- Concrete Factories -----
class WinFactory(GUIFactory):
    def create_button(self) -> Button:
        return WindowsButton()

    def create_checkbox(self) -> Checkbox:
        return WindowsCheckbox()


class MacFactory(GUIFactory):
    def create_button(self) -> Button:
        return MacButton()

    def create_checkbox(self) -> Checkbox:
        return MacCheckbox()


# ----- Client -----
class Application:
    """
    The Client works only with AbstractFactory and AbstractProduct interfaces.
    This allows it to work with any concrete factory/product family.
    """
    def __init__(self, factory: GUIFactory):
        self.factory = factory

    def create_ui(self) -> None:
        button = self.factory.create_button()
        checkbox = self.factory.create_checkbox()

        print(f"  {button.paint()}")
        print(f"  {checkbox.paint()}")

    def render(self) -> None:
        print("\n--- Rendering UI ---")
        self.create_ui()


# ----- Example usage -----
def abstract_factory_example(is_windows: bool = False) -> None:
    """
    Demonstrates the Abstract Factory pattern.

    Args:
        is_windows: If True, uses Windows factory, otherwise Mac factory
    """
    if is_windows:
        factory = WinFactory()
        os_name = "Windows"
    else:
        factory = MacFactory()
        os_name = "Mac"

    print(f"\n--- Abstract Factory: {os_name} GUI ---")
    app = Application(factory)
    app.render()


if __name__ == "__main__":
    abstract_factory_example(is_windows=False)
    print("\n" + "="*50 + "\n")
    abstract_factory_example(is_windows=True)

File: rust/src/a2_abstract_factory.rs
//! # Abstract Factory Pattern
//!
//! The Abstract Factory pattern provides an interface for creating families
//! of related objects without specifying their concrete classes.
//!
//! ## Key Participants
//! - **AbstractFactory**: Declares interface for creating abstract products
//! - **ConcreteFactory**: Implements operations to create concrete products
//! - **AbstractProduct**: Declares interface for a type of product
//! - **ConcreteProduct**: Defines product object to be created by concrete factory
//!
//! ## Benefits
//! - Isolates client classes from concrete classes
//! - Easy to exchange product families
//! - Promotes consistency among products

use std::rc::Rc;

// Abstract Products
pub trait Button {
    fn paint(&self);
}

pub trait Checkbox {
    fn paint(&self);
}

// Concrete Products for Windows
pub struct WindowsButton;

impl Button for WindowsButton {
    fn paint(&self) {
        println!("[Windows] Rendering button");
    }
}

pub struct WindowsCheckbox;

impl Checkbox for WindowsCheckbox {
    fn paint(&self) {
        println!("[Windows] Rendering checkbox");
    }
}

// Concrete Products for Mac
pub struct MacButton;

impl Button for MacButton {
    fn paint(&self) {
        println!("[Mac] Rendering button");
    }
}

pub struct MacCheckbox;

impl Checkbox for MacCheckbox {
    fn paint(&self) {
        println!("[Mac] Rendering checkbox");
    }
}

// Abstract Factory
pub trait GUIFactory {
    fn create_button(&self) -> Rc<dyn Button>;
    fn create_checkbox(&self) -> Rc<dyn Checkbox>;
}

// Concrete Factories
pub struct WindowsFactory;

impl GUIFactory for WindowsFactory {
    fn create_button(&self) -> Rc<dyn Button> {
        Rc::new(WindowsButton)
    }

    fn create_checkbox(&self) -> Rc<dyn Checkbox> {
        Rc::new(WindowsCheckbox)
    }
}

pub struct MacFactory;

impl GUIFactory for MacFactory {
    fn create_button(&self) -> Rc<dyn Button> {
        Rc::new(MacButton)
    }

    fn create_checkbox(&self) -> Rc<dyn Checkbox> {
        Rc::new(MacCheckbox)
    }
}

// Client
pub struct Application {
    factory: Rc<dyn GUIFactory>,
}

impl Application {
    pub fn new(factory: Rc<dyn GUIFactory>) -> Self {
        Self { factory }
    }

    pub fn create_ui(&self) {
        let button = self.factory.create_button();
        let checkbox = self.factory.create_checkbox();

        button.paint();
        checkbox.paint();
    }
}

/// Example demonstrating Abstract Factory pattern
pub fn example() {
    println!("\n--- Abstract Factory Pattern Example ---\n");

    println!("1. Windows Application:");
    let windows_factory = Rc::new(WindowsFactory);
    let windows_app = Application::new(windows_factory);
    windows_app.create_ui();

    println!("\n2. Mac Application:");
    let mac_factory = Rc::new(MacFactory);
    let mac_app = Application::new(mac_factory);
    mac_app.create_ui();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_windows_factory_creates_windows_widgets() {
        let factory = WindowsFactory;
        let button = factory.create_button();
        let checkbox = factory.create_checkbox();

        // Type checking through dynamic dispatch
        button.paint();
        checkbox.paint();
    }

    #[test]
    fn test_mac_factory_creates_mac_widgets() {
        let factory = MacFactory;
        let button = factory.create_button();
        let checkbox = factory.create_checkbox();

        button.paint();
        checkbox.paint();
    }

    #[test]
    fn test_application_with_windows_factory() {
        let factory = Rc::new(WindowsFactory);
        let app = Application::new(factory);
        app.create_ui(); // Should not panic
    }

    #[test]
    fn test_application_with_mac_factory() {
        let factory = Rc::new(MacFactory);
        let app = Application::new(factory);
        app.create_ui(); // Should not panic
    }
}
